Fuzz testing

hulkoba opened the pull request:
# PR description

Relates to #https://github.com/openpgpjs/openpgpjs/issues/1689

Adds fuzz test for
- createCleartextMessage
- createMessage({binary})
- createMessage({text})
- generateKey()
- readKey({ armoredKey })
- readKey({ binaryKey })
- readMessage({ binaryMessage })
- readMessage({ armoredMessage })

## Running a fuzz test

The `test/fuzz` directory contains fuzz targets like for example `createMessageBinary`.

You can run one fuzz target:

```sh
TARGET=createMessageBinary npm run fuzz
```

Notice, that `TARGET` is the name of your fuzz target module.

--- 

## (Don't) running with coverage
Jazzer.js unfortunately does not support coverage for esm tests, yet. 

We've implemented the tests also with coverage, turned everything into cjs, but removed it because it is not super useful.
Because  the imported modules were not read correctly, the result always shows, that the fuzzing function is 100% covered but not the function we are trying to fuzz.

We decided to use esm-tests and wait until jazzer.js adds coverage support for esm.
Then it can be adapted easily, by adding the `--coverage` option to the jazzer command.

hulkoba commented on test/fuzz/readMessageBinary.js:
```diff
@@ -0,0 +1,10 @@
+ * @param { Buffer } inputData
+ */
+export async function fuzz(inputData) {
+  const binaryMessage = new Uint8Array(`-----BEGIN PGP MESSAGE-----${inputData}-----END PGP MESSAGE-----`);
```
I am not sure if this is the best way to test, of if it should be tested

hulkoba commented on test/fuzz/readme.md:
````diff
@@ -0,0 +1,41 @@
+
+You can run this fuzz target with:
+```sh
+npx jazzer test/fuzz/createMessageBinary -- -artifact_prefix=test/fuzz/
````
We can remove the `-artifact_prefix` option and the reports will be put to the root folder. Then we need to adapt the gitignore

hulkoba committed:
test(fuzz): add fuzz tests

hulkoba committed:
chore: install jazzerjs

hulkoba committed:
docs(fuzz): add readme

janl commented on package.json:
```diff
@@ -47,6 +47,8 @@
     "prepare": "npm run build",
     "test": "mocha --timeout 120000 test/unittests.js",
     "test-type-definitions": "node --loader ts-node/esm test/typescript/definitions.ts",
+    "fuzz": "jazzer test/fuzz/$TARGET.cjs -- -artifact_prefix=test/fuzz/reports/",
```
my test output says that using `--sync` might be useful. When I add it here, I get to an OOM relatively quickly (`==45176== ERROR: libFuzzer: out-of-memory (used: 2085Mb; limit: 2048Mb)`), without `--sync` the test goes on “forever”. We should understand what `--sync` exactly does before we decide which way we want to set this.

janl replied:
For example, this run: 

```
> jazzer test/fuzz/$TARGET.cjs --coverage --cov_dir=test/fuzz/coverage --sync -- -artifact_prefix=test/fuzz/reports/ -max_total_time=5

Dictionary: 4 entries
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 2143719647
INFO: Loaded 1 modules   (512 inline 8-bit counters): 512 [0x130050000, 0x130050200), 
INFO: Loaded 1 PC tables (512 PCs): 512 [0x12e800000,0x12e802000), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 1 ft: 1 corp: 1/1b exec/s: 0 rss: 136Mb
#65536	pulse  cov: 1 ft: 1 corp: 1/1b lim: 652 exec/s: 21845 rss: 560Mb
#109377	DONE   cov: 1 ft: 1 corp: 1/1b lim: 1090 exec/s: 18229 rss: 844Mb
Done 109377 runs in 6 second(s)
--------------------|---------|----------|---------|---------|-------------------
File                | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
--------------------|---------|----------|---------|---------|-------------------
All files           |   33.33 |        0 |      20 |   36.36 |                   
 readKeyArmored.cjs |   33.33 |        0 |      20 |   36.36 | 6,14-21           
--------------------|---------|----------|---------|---------|-------------------
```

leads to a coverage report that only covers the `fuzz` export and the dynamic `import` line, but none of the rest. What was tested in the ~1700 executions that never reached the code further down?
<img width="982" alt="Screenshot 2023-12-12 at 15 03 15" src="https://github.com/neighbourhoodie/openpgpjs/assets/11321/e197b80f-a3ae-406d-ae3a-3a3908b07d1f">

hulkoba replied:
the --sync flag only works for synchronous code, which we have now. But we don't use it and don't mention it in the readme.

hulkoba replied:
to prevent the oom happen so fast, you could increase the rss_limit_mb ([Memory usage limit in Mb, default 2048](https://www.llvm.org/docs/LibFuzzer.html#options))

hulkoba replied:
Interesting, for `readKeyArmored` it looks like this

![Screenshot from 2023-12-13 09-21-34](https://github.com/neighbourhoodie/openpgpjs/assets/4191287/b50f69f9-53a5-43dd-9fc7-84f49886b92c)

janl commented on test/fuzz/createCleartextMessage.cjs:
```diff
@@ -0,0 +1,14 @@
+module.exports.fuzz = function(inputData) {
+  import('../initOpenpgp.js').then(openpgp => {
+    const data = new FuzzedDataProvider(inputData);
+    const text = data.bufToPrintableString(inputData, 2, 444, 'utf-8');
```
I’d like to revisit where the hardcoded `444` (or `333` etc below) comes from. It might be worth allowing way larger input strings.

hulkoba replied:
agree

hulkoba replied:
I made it 9000 
`const MAX_MESSAGE_LENGTH = 9000;`

janl replied:
how did you pick 9000? In my output I see `INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes`, so maybe that’s a good limit, as we won’t get any longer input anyway?

hulkoba replied:
it is 4000 now

hulkoba replied:
the maxLength parameter is required and not limited https://github.com/CodeIntelligenceTesting/jazzer.js/blob/main/packages/core/FuzzedDataProvider.ts#L436

janl commented on package.json:
```diff
@@ -47,6 +47,8 @@
     "test": "mocha --timeout 120000 test/unittests.js",
     "test-type-definitions": "node --loader ts-node/esm test/typescript/definitions.ts",
+    "fuzz": "jazzer test/fuzz/$TARGET.cjs -- -artifact_prefix=test/fuzz/reports/",
+    "fuzz-coverage": "jazzer test/fuzz/$TARGET.cjs --coverage --cov_dir=test/fuzz/coverage -- -artifact_prefix=test/fuzz/reports/",
```
the coverage report looks like this:

1x 3390089x 3390088x | /**  * @param { Buffer } inputData  */ module.exports.fuzz = function(inputData) {   import('../initOpenpgp.js').then(openpgp => {     return openpgp.default.createMessage({ binary: new Uint8Array(inputData) });   }); };
-- | --

I had hoped it would actually dive into the library code. The fact that the fuzzing function is 100% covered is not really useful

janl commented on test/fuzz/readKeyArmored.cjs:
```diff
@@ -0,0 +1,26 @@
+  import('../initOpenpgp.js').then(openpgp => {
+    const data = new FuzzedDataProvider(inputData);
+    const fuzzedText = data.consumeString(333, 'utf-8');
+    const armoredKey = `-----BEGIN PGP PRIVATE KEY BLOCK----- ${fuzzedText} -----END PGP PRIVATE KEY BLOCK-----`;
```
what is the significance of using the BEGIN and END blocks? wouldn't it make sense to fuzz the entire input?

hulkoba replied:
When I fuzz the entire input, the errors are caught by the app. I can ignore them, but then the fuzzing does not make sense to me. I commented on this before... absolutely unsure

janl replied:
yeah it could be considered either way:

On the one hand: of course the function will fail any input without the correct BEGIN or END, so we should take them as given and only fuzz inside those markers

On the other hand: the detection of BEGIN and END markers are subject to possible bugs and maybe things can get snuck beyond that check so they should be included in the fuzz input.

_maybe_ we need both? The second run will almost entirely throw errors unless the random input produces the blocks correctly which is very unlikely. but it’d make sure that that part of the fuzzed function is working correctly.

janl replied:
where do you land?

janl commented on test/fuzz/readme.md:
````diff
@@ -0,0 +1,71 @@
+
+You can pass the `-max_total_time` flag to the internal fuzzing engine to stop the fuzzing run after 10 seconds.
+```sh
+npx jazzer test/fuzz/createMessageBinary.cjs -- -max_total_time=10
````
earlier in the file we use `rpm run …` let’s be consistent here

hulkoba replied:
No, we say use npm _or_ npx at the beginning

janl replied:
I thought we agreed on not mentioning npx at all since the project doesn't use it just yet

hulkoba committed:
docs(fuzz): add sync explaination

hulkoba committed:
chore(fuzz): remove coverage

because it is not super useul

hulkoba committed:
test(fuzz): turn everything into esm

since coverage does not work as expected at all, we can use esm. So if jazzer at some point, add esm support for esm, we can easily add it

hulkoba committed:
chore(fuzz): change max message length

hulkoba committed:
chore: create artifacts dir on run

hulkoba committed:
refactor(fuzz): readMessageText -> readMessageArmored

hulkoba committed:
chore(fuzz): import specific function

hulkoba committed:
refactor(fuzz): rename errors: inored -> expected

hulkoba committed:
feat(fuzz): userId.fromObject

hulkoba committed:
fix(fuzz): pass buffered to readBinary funcs
